### PR TITLE
Add missing support for nested JSON objects in arrays

### PR DESCRIPTION
**Pull request by ahmetmircik**
fix attempt for https://github.com/hazelcast/hazelcast/issues/15368

### CODE REVIEW

**Comment by ahmetmircik** [comment]
@mmedenjak thanks fro the findings.
Fixed commit msg and re-wrote fix to not to give NPE, this is because to align behavior with non-array queries which don't throw NPE.

**Comment by mmedenjak** [comment]
Thank you for the update. I just realised that also additional non-existant paths are ignored. For instance, in `MapAggregationJsonTest#testArrayWithNestedField` you can use the path `nestedArray[any].nestedArrayObject.secondary.nestedObjectLongValue.nonExistant` and it should return the same values. Can you check? What do other extractors do in that case?

**Comment by mmedenjak** [comment]
Also, please address comments in a separate commit.

**Comment by ahmetmircik** [comment]
@mmedenjak unfortunately, re-wrote whole fix. A complete review is needed now.

File: hazelcast/src/main/java/com/hazelcast/query/impl/getters/AbstractJsonGetter.java [hunk 1]
```diff
@@ -256,21 +257,31 @@ private MultiResult getMultiValue(JsonParser parser, JsonPathCursor cursor) thro
...
+                            if (pathCursor.getNext() == null) {
+                                addToMultiResult(multiResult, parser);
+                                break;
+                            }
+                        } else {
+                            break;
                         }
-                        parser.nextToken();
                     }
+                    while (true);
```
Comment (mmedenjak) [inline hazelcast/src/main/java/com/hazelcast/query/impl/getters/AbstractJsonGetter.java]: Minor: maybe just move this a line up so it's after the bracket.

File: hazelcast/src/main/java/com/hazelcast/query/impl/getters/AbstractJsonGetter.java [hunk 2]
```diff
@@ -256,21 +257,31 @@ private MultiResult getMultiValue(JsonParser parser, JsonPathCursor cursor) thro
...
+                if (currentToken == JsonToken.START_OBJECT) {
+                    do {
+                        boolean attributeFound = findAttribute(parser, pathCursor);
+                        if (attributeFound
+                                && parser.currentToken() != JsonToken.START_OBJECT
+                                && !pathCursor.hasNext()) {
+                            addToMultiResult(multiResult, parser);
+                            break;
+                        } else if (attributeFound) {
+                            if (pathCursor.getNext() == null) {
```
Comment (mmedenjak) [inline hazelcast/src/main/java/com/hazelcast/query/impl/getters/AbstractJsonGetter.java]: What is this branch for? I tested it by changing the path in `MapAggregationJsonTest#testArrayWithNestedField` to `nestedArray[any].nestedArrayObject.secondary` but I only managed to get the original exception that this PR is fixing:
```
16:30:26,838 ERROR |testArrayWithNestedField[inMemoryFormat: BINARY, metadataPolicy: CREATE_ON_UPDATE]| - [QueryPartitionOperation] hz.friendly_taussig.partition-operation.thread-1 - [127.0.0.1]:5703 [dev] [4.0-SNAPSHOT] com.hazelcast.internal.json.NonTerminalJsonValue cannot be cast to java.lang.Comparable
java.lang.ClassCastException: com.hazelcast.internal.json.NonTerminalJsonValue cannot be cast to java.lang.Comparable
	at com.hazelcast.aggregation.impl.MaxAggregator.accumulateExtracted(MaxAggregator.java:27) ~[classes/:?]
	at com.hazelcast.aggregation.impl.AbstractAggregator.accumulate(AbstractAggregator.java:75) ~[classes/:?]
	at com.hazelcast.map.impl.query.AggregationResult.add(AggregationResult.java:88) ~[classes/:?]
	at com.hazelcast.map.impl.query.PartitionScanRunner.run(PartitionScanRunner.java:108) ~[classes/:?]
	at com.hazelcast.map.impl.query.CallerRunsPartitionScanExecutor.execute(CallerRunsPartitionScanExecutor.java:42) ~[classes/:?]
	at com.hazelcast.map.impl.query.QueryRunner.runPartitionIndexOrPartitionScanQueryOnGivenOwnedPartition(QueryRunner.java:196) ~[classes/:?]
	at com.hazelcast.map.impl.query.QueryPartitionOperation.runInternal(QueryPartitionOperation.java:46) ~[classes/:?]
	at com.hazelcast.map.impl.operation.MapOperation.run(MapOperation.java:109) [classes/:?]
	at com.hazelcast.spi.impl.operationservice.Operation.call(Operation.java:171) ~[classes/:?]
	at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.call(OperationRunnerImpl.java:210) [classes/:?]
	at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.run(OperationRunnerImpl.java:199) [classes/:?]
	at com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.process(OperationThread.java:147) [classes/:?]
	at com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.process(OperationThread.java:125) [classes/:?]
	at com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.run(OperationThread.java:110) [classes/:?]
```

File: hazelcast/src/main/java/com/hazelcast/query/impl/getters/AbstractJsonGetter.java (hunk 1 above)
Comment (mmedenjak) [inline hazelcast/src/main/java/com/hazelcast/query/impl/getters/AbstractJsonGetter.java]: Minor: maybe just move this line above after the bracket.

File: hazelcast/src/main/java/com/hazelcast/query/impl/getters/AbstractJsonGetter.java (hunk 2 above)
Comment (ahmetmircik) [reply hazelcast/src/main/java/com/hazelcast/query/impl/getters/AbstractJsonGetter.java]: now it shouldn't give exception + improved the code.

**Pull request merged**